Paralogues.pm

```diff
@@ -0,0 +1,705 @@
+    die "Translation $protein_id not cached; avoid using --offline to allow to connect to database\n";
+  }
+
+  my $species = $config->species;
+  my $reg     = $config->registry;
+  $self->{ta} ||= $reg->get_adaptor($species, 'core', 'translation');
+
+  my $transcript = $self->{ta}->fetch_by_stable_id($protein_id)->transcript;
+  $self->{_cache}->{$protein_id} = $transcript;
+  return $transcript;
+}
+
+## PLUGIN ----------------------------------------------------------------------
+
+sub new {
+  my $class = shift;  
+  my $self = $class->SUPER::new(@_);
+
+  $self->expand_left(0);
+  $self->expand_right(0);
+  $self->get_user_params();
+
+  my $params = $self->params_to_hash();
+  my $config = $self->{config};
+
+  # Check for custom VCF
+  my $vcf = $params->{vcf};
+  if (defined $vcf) {
+    $self->{vcf} = $vcf;
+    $self->add_file($vcf);
+
+    # get INFO fields names from VCF
+    my $vcf_file = Bio::EnsEMBL::IO::Parser::VCF4Tabix->open($vcf);
+    my $info = $vcf_file->get_metadata_by_pragma('INFO');
+    my $info_ids = [ map { $_->{ID} } @$info ];
+
+    if ( !defined($params->{cols}) ) {
+      $self->{cols} = $info_ids;
+    } else {
+      my @cols = split(/:/, $params->{cols});
+
+      # check if the selected INFO fields exist
+      $self->{cols} = [];
+      my @invalid_cols;
+      for my $col (@cols) {
+        if ( grep { $_ eq $col } @$info_ids ) {
+          push(@{$self->{cols}}, $col);
+        } else {
+          push(@invalid_cols, $col);
+        }
+      }
+
+      my $filename = basename $vcf;
+      die "ERROR: input INFO fields not found in $filename. Available INFO fields are:\n" .
+        join(", ", @$info_ids)."\n" unless @{ $self->{cols} };
+
+      warn "WARNING: the following INFO fields were not found in $filename and were ignored: ",
+        join(", ", @invalid_cols), "\n" if @invalid_cols;
+    }
+  } elsif ($config->{offline}) {
+    die("ERROR: Cannot fetch Ensembl variants in offline mode; please define vcf argument in the Paralogues plugin\n");
+  }
+
+  # Check if paralogue annotation should be downloaded
+  $self->{remote} = defined $params->{mode} && $params->{mode} eq 'remote';
+  return $self if $self->{remote};
+
+  # Generate paralogue annotation
+  my $annot = defined $params->{paralogues} ? $params->{paralogues} : $self->_prepare_filename;
+
+  my $dir = $params->{dir};
+  if (defined $dir) {
+    die "ERROR: directory $dir not found" unless -e -d $dir;
+    $annot = File::Spec->catfile( $dir, $annot );
+  }
+
+  $self->_generate_paralogue_annotation($annot) unless (-e $annot || -e "$annot.lock");
+  $self->{paralogues} = $annot;
+  return $self;
+}
+
+sub feature_types {
+  return ['Feature'];
+}
+
+sub get_header_info {
+  my $self = shift;
+
+  my @fields;
+  if ( defined $self->{vcf} ) {
+    @fields = (
+      'identifier',
+      'sequence',
+      'start',
+      'alleles'
+    );
+    push @fields, @{ $self->{cols} };
+  } else {
+    @fields = (
```

Comment:
Defaulting to only variant name, alleles & clinical significance would be enough.
Being able to report paralogous gene symbol and match quality would be useful, so people can instantly spot gene families and think about relevance.

Reply:
The `fields` argument handling was improved in 7d4a7f2397690f651f40346a2d9541a859b5fa8f and 010c46d:
- Defaults to variant identifier, alleles and clinical significance
- Allow to report new fields:
    - Ensembl API: `gene_symbol`
    - VCF file: `quality` and `filter` columns
    - Both: `perc_cov` and `perc_pos`